experimental/PropertyDDS/packages/property-changeset/src/changeset_operations/array.ts

```diff
@@ -5,73 +5,115 @@
 /**
  * @fileoverview Helper functions and classes to work with array ChangeSets
  */
+// @ts-ignore
+import { ConsoleUtils, constants } from "@fluid-experimental/property-common";
+import _ from "lodash";
+import { ApplyChangeSetOptions, ConflictInfo, SerializedChangeSet } from "../changeset";
+import { TypeIdHelper } from "../helpers/typeidHelper";
+import { ArrayChangeSetIterator, arrayInsertList, arrayModifyList, arrayRemoveList, GenericOperation, InsertOperation, ModifyOperation, NoneNOPOperation, NOPOperation, RemoveOperation } from "./arrayChangesetIterator";
+import { ConflictType } from "./changesetConflictTypes";
 
-import _ from "lodash"
-import {constants, ConsoleUtils} from "@fluid-experimental/property-common";
-import { isPrimitiveType as _isPrimitiveType } from "../helpers/typeidHelper";
-import { INSERTED_ENTRY_WITH_SAME_KEY,
-     ENTRY_MODIFIED_AFTER_REMOVE, COLLIDING_SET, REMOVE_AFTER_MODIFY } from "./changesetConflictTypes";
-import ArrayChangeSetIterator, { types } from "./arrayChangesetIterator";
+const { MSG } = constants;
+const { isPrimitiveType } = TypeIdHelper;
 
-const MSG = constants.MSG;
 
 /**
- * @namespace property-changeset.ChangeSetOperations
- * @alias property-changeset.ChangeSetOperations
- * Helper functions and classes to perform operations on ChangeSets
+ * The range combinations of two change sets (A and B)
+ * This can either be complete operations, parts of complete operations or overlapping segments
+ * @enum number
  */
+enum ArrayChangeSetRangeType {
+    completeA, // a complete operation of change set A
+    completeB, // a complete operation of change set B
+    partOfA, // a partial operation of change set A
+    partOfB, // a partial operation of change set B
+    completeApartOfB, // a complete operation of change set A overlapping with a partial operation of change set B
+    completeBpartOfA, // a complete operation of change set B overlapping with a partial operation of change set A
+    completeAcompleteB, // a complete operation of change set A overlapping a complete operation of change set B
+    partOfApartOfB, // a partial operation of change set A, a partial operation of change set B
+};
+
+
+interface SegmentType<T=GenericOperation, K=GenericOperation, L=GenericOperation> {
+    begin?: number;
+    op?: T;
+    flag?: ArrayChangeSetRangeType;
+    opA?: K;
+    opB?: L;
+    removeInsertOperationA?: arrayRemoveList|arrayInsertList;
+
+}
 
 /**
- * @namespace property-changeset.ChangeSetOperations.ArrayOperations
- * @alias property-changeset.ChangeSetOperations.ArrayOperations
- * Helper functions and classes to perform operations on array ChangeSets
+ * A range of an array operation
  */
+export interface OperationRangeDescription<T=GenericOperation> {
+    opA?: any;
```

Comment:
yes, probably, but it will break lot of places in the code. Because we have a lot of places where we initialize properties partially, and then it will complain on all these places.

Reply:
I think we should re-iterate on those after we introduce the comprehensive ChangeSet type.